### PR TITLE
Execution tests batch for CEK

### PR DESCRIPTION
**Pull request by 0xd34df00d**
PR checklist:

* [x] Test coverage for the proposed changes
* [x] PR description contains example output from repl interaction or a snippet from unit test output
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/kadena-io/pact-core/blob/master/CHANGELOG.md)

Additionally, please justify why you should or should not do the following:

* [ ] Confirm replay/back compat
* [ ] Benchmark regressions
* [ ] (For Kadena engineers) Run integration-tests against a Chainweb built with this version of Pact

### CODE REVIEW

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 1]
```diff
@@ -392,7 +400,7 @@ applyPact i pc ps cont handler cenv nested = useEvalState esDefPactExec >>= \cas
 
       -- Check we try to apply the correct pact Step
       unless (ps ^. psStep < nSteps) $
-        throwExecutionError i (DefPactStepNotFound ps nSteps)
+        throwExecutionError i (DefPactStepNotFound ps nSteps)   -- TODO TODO how to trigger this?
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact-core/Pact/Core/IR/Eval/CEK.hs:403]: This one cannot be triggered at all. Checked with @0xd34df00d in a call.

File: pact-core/Pact/Core/Environment/Utils.hs [hunk 2]
```diff
@@ -167,6 +139,7 @@ getModuleMember info pdb (QualifiedName qn mn) = do
     Nothing -> do
       let fqn = FullyQualifiedName mn qn (_mHash md)
       throwExecutionError info (NameNotInScope fqn)
+      -- TODO ^ failed to find a test
```
Comment (jmcardon) [inline pact-core/Pact/Core/Environment/Utils.hs:142]:
```
(module m g
  (defcap g () true)

  (defun f () 1)

  (defschema georg-schema elem:guard)
  (deftable georg:{georg-schema})
  )
(create-table georg)
(insert georg "k" {"elem":(create-user-guard (f))})

(module m g
  (defcap g () true)
  (defschema georg-schema elem:guard)
  (deftable georg:{georg-schema})
  )

(enforce-guard (at "elem" (read georg "k")))
```

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 3]
```diff
@@ -161,8 +161,13 @@ evaluateTerm cont handler env (Var n info)  = do
           returnCEKValue cont handler (VClosure (CT clo))
         Just d ->
           throwExecutionError info (InvalidDefKind (defKind d) "in var position")
+          -- TODO ^ shall this be an invariant failure?
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:164]: Agree, invariant failure.

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 4]
```diff
@@ -161,8 +161,13 @@ evaluateTerm cont handler env (Var n info)  = do
...
+          -- apparently this can never happen because `renameTerm` checks for variable top-level variable occurrences,
+          -- erroring out with `InvalidDefInTermVariable` if it's violated.
+          -- And, the execution can't change that invariant once it's established during desugar.
         Nothing ->
           throwExecutionError info (NameNotInScope (FullyQualifiedName mname (_nName n) mh))
+          -- TODO ^ ditto, `renameTerm` apparently always fails in this case as well
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:170]: yes, but I want info about the name not in scope

Comment (jmcardon) [reply pact-core/Pact/Core/IR/Eval/CEK.hs:170]: Maybe just the show instance is fine? Or just `renderQualifiedName` or something

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 5]
```diff
@@ -260,6 +265,7 @@ evaluateTerm cont handler env (CapabilityForm cf info) = do
   case cf of
     WithCapability rawCap body -> do
       enforceNotWithinDefcap info env "with-capability"
+      -- TODO this seems to be an invariant failure at this point since desugar takes care of this check
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:268]: This one is because of Prod!!!

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 6]
```diff
@@ -326,6 +332,8 @@ mkDefunClosure d mn e = case _dfunTerm d of
     pure (Closure (_dfunName d) mn NullaryClosure 0 body (_dfunRType d) e i)
   _ ->
     throwExecutionError (_dfunInfo d) (DefIsNotClosure (_dfunName d))
+    -- TODO ^ apparently invariant failure, since the parser ensures defun has a form of `(defun ( args ) body)`,
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:335]: Agreed. Also invariant.

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 7]
```diff
@@ -535,13 +543,13 @@ resumePact
   -> Maybe DefPactExec
   -> m (CEKEvalResult step b i m)
 resumePact i cont handler env crossChainContinuation = viewEvalEnv eeDefPactStep >>= \case
-  Nothing -> throwExecutionError i DefPactStepNotInEnvironment
+  Nothing -> throwExecutionError i DefPactStepNotInEnvironment    -- <- TODO apparently can't happen, `continuePact` ensures that
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:546]: Can't be triggered from REPL either, but a caller on chainweb's side could've done it.

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 8]
```diff
@@ -553,27 +561,27 @@ resumePact i cont handler env crossChainContinuation = viewEvalEnv eeDefPactStep
...
-          throwExecutionError i (CCDefPactContinuationError ps ccExec dbExec)
+          throwExecutionError i (CCDefPactContinuationError ps ccExec dbExec)    -- TODO TODO how to trigger this?
 
         resumeDefPactExec ccExec
       where
         --resumeDefPactExec :: CEKEval step b i m, MonadEval b i m => DefPactExec -> m (CEKEvalResult step b i m)
         resumeDefPactExec pe = do
           when (_psDefPactId ps /= _peDefPactId pe) $
-            throwExecutionError i (DefPactIdMissmatch (_psDefPactId ps) (_peDefPactId pe))
+            throwExecutionError i (DefPactIdMismatch (_psDefPactId ps) (_peDefPactId pe))    -- TODO TODO how to trigger this?
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:575]: Also an API one.

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 9]
```diff
@@ -607,6 +615,7 @@ nameToFQN info env (Name n nk) = case nk of
       md <- getModule info (view cePactDb env) (_mrModule mr)
       pure (FullyQualifiedName (_mrModule mr) dArg (_mHash md))
     Just _ -> throwExecutionError info (DynNameIsNotModRef dArg)
+    -- TODO this ^ is supposedly caught by the typechecker, so it's an invariant error now
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:618]: OOH great catch! We're missing the runtime checks on modrefs.

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 10]
```diff
@@ -933,6 +942,7 @@ enforceNotWithinDefcap
   -> m ()
 enforceNotWithinDefcap info env form =
   when (_ceInCap env) $ throwExecutionError info (FormIllegalWithinDefcap form)
+  -- TODO this seems to be an invariant failure at the point where this is called, since desugar takes care of this
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:945]: We're leaving this one in for prod compat.

File: pact-core/Pact/Core/IR/Eval/CEK.hs [hunk 11]
```diff
@@ -1028,6 +1038,7 @@ installCap info _env (CapToken fqn args) autonomous = do
     DefManaged m -> case m of
       DefManagedMeta (paramIx,_) (FQName fqnMgr) -> do
         managedParam <- maybe (throwExecutionError info (InvalidManagedCap fqn)) pure (args ^? ix paramIx)
+        -- TODO ^ OOB index is invariant failure?
```
Comment (jmcardon) [inline pact-core/Pact/Core/IR/Eval/CEK.hs:1041]: Correct. This is definitely an invariant failure, this is done in desugar.

**Comment by 0xd34df00d** [comment]
Merging this now as per offline discussion. I'll do typed invariant failures as a separate PR.

**Pull request merged**